pkg/expression/builtin_compare.go

```diff
@@ -141,6 +141,10 @@ func (c *coalesceFunctionClass) getFunction(ctx sessionctx.Context, args []Expre
 		return nil, err
 	}
 
+	if resultFieldType.GetType() == mysql.TypeDate {
+		bf.setDecimalAndFlenForDate()
+	}
```

Comment:
I'm ok with this fix for this specific issue.
But I wonder why there is no `ETDate`, which makes this type deduction wrong, which cause we have to add this kind of code for every similar situation.

Reply:
> Why not just set resultFieldType to bf.tp? For control functions(case when/ifnull), it use this way to avoid this problem.

The type of resultFieldType is `EvalType` and `EvalType` doesn't have `Date` type.

Reply:
I think `resultEvalType` is the `EvalType`, `resultFieldType` is field type.

Reply:
> I think `resultEvalType` is the `EvalType`, `resultFieldType` is field type.

okk, I will change it.

Reply:
> I think `resultEvalType` is the `EvalType`, `resultFieldType` is field type.

From the failure of ut, we can see that when sql is `select coalesce(c_int_d, c_datetime) from xxx;` the test will fail. So, we can't directly set resultFieldType to bf.tp.